File: Countdown.scala
class Countdown(numSeq: Vector[Int]) {

  type State = Int
  val initialState = 0


  trait Op {
    def isApplicable(state: State): Boolean

    def apply(state: State): State
  }

  case class Add(v: Int) extends Op {
    def isApplicable(state: State) = true

    def apply(state: State) = state + v
  }

  case class Sub(v: Int) extends Op {
    def isApplicable(state: State) = state > v

    def apply(state: State) = state - v
  }

  case class Mul(v: Int) extends Op {
    def isApplicable(state: State) = true

    def apply(state: State) = state * v
  }

  case class Div(v: Int) extends Op {
    def isApplicable(state: State) = state % v == 0

    def apply(state: State) = state / v
  }

  val ops = Vector(Add, Sub, Mul, Div)

  val moves =
    for {
      v <- numSeq
      op <- ops
    }
      yield op(v)

  class Path(history: List[Op], val endState: State) {
    def extend(op: Op) = {
      new Path(op :: history, op apply endState)
    }

    override def toString = (history.reverse mkString " ") + "--> " + endState
  }

  val initialPath = new Path(Nil, initialState)
  val visitedStates = scala.collection.mutable.Set[State](initialState)

  def from(paths: Set[Path]): Stream[Set[Path]] =
    if (paths.isEmpty) Stream.empty
    else {
      val more = for {
        path <- paths
        validMoves = moves.filter(m => m.isApplicable(path.endState))
        next <- validMoves map path.extend
        if !(visitedStates contains next.endState)
      } yield next
      visitedStates ++= more.map(m => m.endState)
      more #:: from(more)
    }

  val pathSets = Set(initialPath) #:: from(Set(initialPath))

  def solutions(target: Int): Stream[Path] = {
    for {
      pathSet <- pathSets
      path <- pathSet
      if path.endState == target
    } yield path
  }

}

File: countdown.hs
import qualified Data.Set as Set

data Op = Add Int | Sub Int | Mul Int | Div Int deriving (Eq, Show, Ord)

apply :: Op -> Int -> Int
apply op state = case op of
            Add v -> state + v
            Sub v -> state - v
            Mul v -> state * v
            Div v -> state `quot` v

isApplicable :: Op -> Int -> Bool            
isApplicable op state = case op of
            Add v -> True
            Sub v -> state >= v
            Mul v -> True
            Div v -> state `mod` v == 0

data Path = Path {history :: [Op], endState :: Int} deriving (Eq, Ord)

instance Show Path where
    show (Path history endSate) = show (reverse history) 

extend :: Path -> Op -> Path
extend path op = Path { history = op : history path, endState = apply op (endState path)}  

ops = [Add, Sub, Mul, Div]

getMoves :: [Int] -> [Op]
getMoves numSeq = [op v | v <- numSeq, op <- ops]

initialState = 0

initialPath = Path {history=[], endState=initialState}

from :: [Int] -> Set.Set(Path) -> Set.Set(Int) -> [Set.Set(Path)]
from numSeq pathset visited = if null pathset then [] else
            let more = Set.fromList(do 
                path <- (Set.elems pathset)
                let moves = getMoves numSeq
                let validMoves = filter (\m -> isApplicable m (endState path) ) moves
                let next = map (extend path) validMoves
                filter (\p -> Set.notMember (endState p) visited ) next)
            in more : from numSeq more (Set.union visited (Set.map endState more) )    

countdown :: [Int] -> Int -> Path
countdown numSeq target = 
    let pathsets = Set.singleton initialPath : from numSeq (Set.singleton initialPath) (Set.singleton initialState)
    in  head [path | pathset <- pathsets, path <- Set.elems pathset, endState path == target]